protocol.md

```diff
@@ -0,0 +1,249 @@
+
+The goal of this protocol is to allow people running Ethereum nodes with Whisper service enabled to exchange messages that are encrypted end-to-end in a way that guarantees darkness to some extent (messages are broadcasted through the network rather than sent directly or using a known path).
+
+It's important to notice that messages are not limited to be text messages only. They can also have special meaning depending on the client implementation. For example, in the current implementation, there are message which informs about Eth requests or current user's location.
+
+This protocol consist of three components:
+* payload specification
+* Whisper adapter
+* Offline messaging.
+
+The payload specification describes how the messages are encoded and decoded and what each fields mean. This is required to properly interpret messages by the client.
+
+Whisper adapter specifies interaction with the Whiper service with regards to keys management, configuration and some metadata (like topic) required to properly process and encrypt/decrypt messages.
+
+Offline messaging describes how the protocol handles delivering messages when one or more participants were offline and the messages expired in the network.
+
+The protocol does not specify things like peers discovery, however, some notes and the current implementation will be described in an appendix.
+
+# Payload
+
+Payload is encrypted and decrypted using [transit format](https://github.com/cognitect/transit-format). It is a text-based format so that each decrypted message's payload can be easily read.
+
+Payload contains the following fields:
+* text `string`
+* content type `enum { text/plain }`
+* message type `enum { public-group-user-message, user-message }`
+* clock `int64`
+* timestamp `int64`
+* content `struct { chat-id string, text string }`
+
+Example of a valid message:
+
+```
+["~#c4",["abc123","text/plain","~:public-group-user-message",154593077368201,1545930773682,["^ ","~:chat-id","testing-adamb","~:text","abc123"]]]
+```
+
+As you can see, the message is an array and each index value has its meaning:
+* 0: `c4` is a decoder handler identificator for the current payload format. Identificators allow to register handler for many different types of messages
+* 1: array which items correspond to the described payload fields above
+
+For more details regarding serialization and deserialization please consult [transit format](https://github.com/cognitect/transit-format) specification.
+
+## Content types
+
+Here is a list of supported content types:
+* `text/plain` identifies a message which content is a plain text
+
+TODO: are there other types? How is it useful? Sending Eth requests?
+
+## Message types
+
+Here is a list of supported message types:
+* `public-group-user-message` is a message to the public group
+* `user-message` is a private message.
+
+TODO: how is it useful? What are other supported message types?
```

Comment:
@yenda @rasom can you guys answer this?

Reply:
There is 3 messages types allowed in the schemas
```clojure
(spec/def ::message-type #{:group-user-message :public-group-user-message :user-message})
```

it is useful to distinguish public and private group and 1-1 chats. There is probably better solutions but it would break compatibility